Intel_adsp_ace1x: Meteorlake board upstream

aborisovich opened the pull request:
Publishing new ACE1X SoC Meteorlake board.
This set of commits is related to board and hardware definitions.
Note for reviewers:

- This PR is a "first half" of the upstream process. It contains zephyr board, soc and hardware characteristics.
- Second part will be published later - it will contain drivers code.

nashif committed:
intel_adsp: add the board definition for the Meteorlake

Board definition for the meteorlake platform.
Enable intel_adsp_ace15_mtpm in the sof sample.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>
Co-authored-by: Michal Wasko <michal.wasko@intel.com>
Co-authored-by: Konrad Leszczynski <konrad.leszczynski@intel.com>

nashif committed:
intel_adsp: add a new series to support Meteorlake

Meteorlake support as part of the Intel ADSP family.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>
Co-authored-by: Michal Wasko <michal.wasko@intel.com>
Co-authored-by: Konrad Leszczynski <konrad.leszczynski@intel.com>
Co-authored-by: Rafal Redzimski <rafal.f.redzimski@intel.com>
Co-authored-by: Enjia Mai <enjia.mai@intel.com>
Co-authored-by: Flavio Ceolin <flavio.ceolin@intel.com>
Co-authored-by: Tomasz Leman <tomasz.m.leman@intel.com>
Co-authored-by: Bonislawski Adrian <adrian.bonislawski@intel.com>
Co-authored-by: Serhiy Katsyuba <serhiy.katsyuba@intel.com>
Co-authored-by: Andrey Borisovich <andrey.borisovich@intel.com>

rfredzim committed:
ace: multiprocessing: enable secondary core boot through ROM

Add ace/mtl rom flags definitions.
Set the flags in battr to indicate to rom that it should execute
secondary core procedure.

Signed-off-by: Rafal Redzimski <rafal.f.redzimski@intel.com>
Co-authored-by: Andrey Borisovich <andrey.borisovich@intel.com>

rfredzim committed:
ace: multiprocessing: add polling for core powered up confirmation

Poll for CPA bit by which HW confirms that the core has been powered up.

Signed-off-by: Rafal Redzimski <rafal.f.redzimski@intel.com>
Co-authored-by: Tomasz Leman <tomasz.m.leman@intel.com>
Co-authored-by: Andrey Borisovich <andrey.borisovich@intel.com>

tmleman committed:
soc: ace_v1x: soc_adsp_halt_cpu implementation

This patch add soc_adsp_halt_cpu implementation for ace. Function disable
power and check CPA status to report success. Function should be used only
for secondary cores and can be executed only from primary core.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman committed:
soc: ace_v1x: pm_state_set implementation

This patch implements pm_state_set function for ACE platforms.

This is initial implementation and only includes the basic handling of
PM_STATE_SOFT_OFF.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman committed:
soc: ace_v1x: pm_state_exit_post_ops implementation

Implementation of the function used during the exit from the lower power
states.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

Commit:
dts: align CONFIG's with ace hardware

Align CONFIG_HEAP_MEM_POOL_SIZE & CONFIG_DAI with ace hardware.

Signed-off-by: Arsen Eloglian <ArsenX.Eloglian@intel.com>

mwasko commented:
@aborisovich there is a lot of commits that require squash in this PR. All the renames, remove commits are irrelevant information on the upstream. Similar apply to some minor changes related to board files and ACE 1.x SoC.

mwasko reviewed (dismissed):
Please clean up commits and squash irrelevant ones.

mwasko reviewed (commented):
@aborisovich will the Meteorlake board compile without the configuration file? If not then it should be part of the Meteorlake board commit.

kv2019i reviewed (changes requested):
Thanks @aborisovich ! Overall looks good. Aside a few minor issues, bigger issues are shared definitions with cAVSxx (either clean up the old one, or use common header for common parts), and what we do with xtensa/config.

aborisovich commented on soc/xtensa/intel_adsp/ace_v1x/cavs-ipc-regs.h:
```diff
@@ -0,0 +1,130 @@
+	} agents[2];
+};
+
+/* FIXME: devicetree */
```
Yes. IPC/IDC is work in progress.

aborisovich commented on soc/xtensa/intel_adsp/ace_v1x/xtensa/config/core-isa.h:
```diff
@@ -0,0 +1,586 @@
+/* 
+ * xtensa/config/core-isa.h -- HAL definitions that are dependent on Xtensa
```
Thanks, removed xtensa headers

aborisovich commented on soc/xtensa/intel_adsp/ace_v1x/multiprocessing.c:
```diff
@@ -0,0 +1,68 @@
+
+	__asm__ volatile("rsr %0, PRID" : "=r"(prid));
+	return prid;
+}
```
Lets do this after upstream

nashif replied:
https://github.com/zephyrproject-rtos/zephyr/pull/46659

aborisovich commented on soc/xtensa/intel_adsp/ace_v1x/cavs-ipc-regs.h:
```diff
@@ -0,0 +1,130 @@
+#define CAVS_IPC_CTL_TBIE BIT(0)
+#define CAVS_IPC_CTL_IDIE BIT(1) /* yesthatiswhattheyreallynamedit */
+
+/* MTL provides an array of six IPC endpoints to be used for
```
I was lacking some commits, now ace-ipc-regs.h is proper one and has all the content

aborisovich commented:
What can I do with soc/xtensa/intel_adsp/ace_v1x/_soc_inthandlers.h file that is missing copyrights?
It is machine generated file. Can I somehow add exclusion to license checker?
@nashif please have a look.

kv2019i commented on soc/xtensa/intel_adsp/ace_v1x/xtensa/config/tie-asm.h:
```diff
@@ -0,0 +1,24 @@
+/* Copyright (c) 2021 Intel Corporation
```
This single xtensa-config file is still include. On purpose?

aborisovich replied:
Yeah, check comments content of the header.

kv2019i commented on soc/xtensa/intel_adsp/ace_v1x/ace-ipc-regs.h:
```diff
@@ -1,8 +1,8 @@
-/* Copyright (c) 2022 Intel Corporation
+/* Copyright (c) 2021 Intel Corporation
```
I think changes to soc/xtensa/ should be squashed to the first patch in the series that adds the ace_v1x soc and this commit should only add the board.

kv2019i commented on soc/xtensa/intel_adsp/ace_v1x/CMakeLists.txt:
```diff
@@ -4,6 +4,7 @@
 # SPDX-License-Identifier: Apache-2.0
```
I'll let @nashif  and @mwasko comment, but these could be also folder to the first patch.

kv2019i commented on soc/xtensa/intel_adsp/ace_v1x/soc.c:
```diff
@@ -11,10 +11,54 @@
 #include <init.h>
```
Ditto for this, could be folder into a single commit that adds the soc.

kv2019i commented on soc/xtensa/intel_adsp/common/CMakeLists.txt:
```diff
@@ -100,3 +100,5 @@ add_custom_target(
       ${CMAKE_BINARY_DIR}/zephyr/main.mod 2>${NULL_FILE}
 )
+
+add_subdirectory(ace)
```
Hmm, this is a bit strange. Why have a "ace" folder under "common" when there is a separate folder for this hardware already?

aborisovich replied:
Yes it is very strange and obviously wrong.
I am just not sure how much refactoring I can push here.
And it could probably affect also SOF upstream.
Maybe lets skip it for now?

kv2019i reviewed (changes requested):
Looking better. Request-changes for comment on first two patches. Remaining comments are not blocking. I'll let @nashif  and @mwasko to chime in how much the patches should be rolled up in this initial PR. Stuff like "intel_adsp: meteorlake: call soc_init" seems like something that should be squashed to a single commit that adds a minimal but functional SoC definition. Later patches that add PM and multicore functionalities may be beneficial to keep as separate patches. 

@aborisovich Also good to check whether the order is correct, so bisect is not broken. E.g. patch that adds a new board target, should come after all the dependencies.

andyross commented on soc/xtensa/intel_adsp/common/ace/power_down.S:
```diff
@@ -0,0 +1,159 @@
+#define u32_ipc_response_mask        a14
+#define pfl_reg                      a15
+
+ace_power_down:
```
To repeat from earlier: I can't see how we can reach ace_power_down() from the code as contributed.  Is this not dead code?  How do we exercise?

tmleman replied:
@andyross 
file: multiprocessing.c
function: pm_state_set
line: 165

If your concerns are about that it will not be executed without **\"_CONFIG_PM=y_\"**, thats true. Unfortunately, no one has raised such objections before.

ceolin replied:
This is optional and applications enable / disable it as they want. The majority of boards supporting PM don't enable it by default.

andyross commented on dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:
```diff
@@ -0,0 +1,126 @@
+		off: off {
+			compatible = "zephyr,power-state";
+			power-state-name = "soft-off";
+			min-residency-us = <2147483647>;
```
A min residency of INT_MAX is more than half an hour.  Not sure where this comes from, and I'll buy that this is probably ignored by existing code, but that actual number is almost certainly a lie.  This doesn't seem to be a pattern used anywhere else in Zephyr, and I don't see any code trying to interpret INT_MAX specially here.  Can we put a real number in here that matches the design intent of the field?

tmleman replied:
@andyross the reason why this state has such a strange value is explained in the comments above the definition. FW will enter this state only on the request from the SW. There is of course permanent lock on application side but I wanted to be sure that this state will not be entered by mistake.

During the first review, it was proposed to remove this state from dts, since it can be entered only by using **_pm_state_force_**. But in my opinion, it would be unfair for a person who will come into contact with this code in some time and see that the processor enters the state that according to board definition is not even supported.

Do you have a suggestion what would be better than the maximum value?

andyross replied:
If the PM scheduling code tries to put you in a state you can't support, don't we have an API for that already by returning the actual state or an error code or something?  I mean, at the end of the day the PM integration is clearly empowered to implement rules like "don't power the core off unless we've received a power-down message from the host" regardless of what the PM layer wants.  So I don't see why we need to lie to the PM layer, I guess.

(For clarity: this comment isn't driving my -1.  It just seems clumsy to me and I suspect there's a misunderstanding of the PM architecture behind it, though I'm not the expert to make that call.)

tmleman replied:
@andyross What value would not be a lie, since there is no number that apply to this state? And to be clear if you misunderstood me before. This is valid state that is supported by each core and will be entered by them.

Our custom pm policy will guarantee that this state will not be entered any other way than on host request. The only other solution I see is to set min-residency-us equal 0 and remove this state from cpu-power-states list (or maybe just set it to 0).

andyross commented on dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:
```diff
@@ -0,0 +1,126 @@
+	sram1: memory@a0000000 {
+		device_type = "memory";
+		compatible = "mmio-sram";
+		reg = <0xa0000000 DT_SIZE_K(60)>;
```
LP-SRAM is 60k?  Note 64k?  This looks like a typo to me.  If it's not, it probably deserves a comment explaining why there's a mismatch with the same SRAM block from cAVS.

(I have a vague sense the answer is going to be "the block is 64k, but there are 4k reserved for this special purpose".  In which case the review comment becomes "Devicetree is for hardware config, software control over memory spaces belongs elsewhere, probably in the linker script as configured by kconfig.")

aborisovich replied:
You are right, this is L2 ULP Memory and is equal 64KB,

aborisovich replied:
You were right, this is L2 ULP Memory 64KB size.

andyross commented on dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:
```diff
@@ -0,0 +1,126 @@
+		reg = <0xa0000000 DT_SIZE_K(60)>;
+	};
+
+	reserverd_mem: reserved-memory {
```
This is a new feature and I'll admit I'm not 100% sure of how it's supposed to be used (@carlocaione might have input).  But this seems wrong to me.  This is for memory that is provided by the platform but not available to "regular" Zephyr usage, so it has its own driver interface.

In fact IMR isn't "reserved" to the linker script.  Zephyr can place regular C data and code there with __imr/__imrdata attributes.  This probably wants to be a regular RAM block symmetric with the existing ones for hp/lp-sram.

carlocaione replied:
`reserved-memory` support as compatible was removed with #45030.

andyross replied:
Ah, OK.  See, this is why the ability to pull in upstream subsystem authors during review is valuable.  I saw this and actually thought it was the new feature and not the old one because I haven't been paying enough attention. :)

andyross commented on dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:
```diff
@@ -0,0 +1,126 @@
+			interrupts = <4 0 0>;
+			num-irqs = <28>;
+			interrupt-parent = <&core_intc>;
+			label = "ACE_0";
```
What uses that label?  ACE has its own driver for this hardware array, and it doesn't seem to care.

aborisovich replied:
Removing useless label.

andyross commented on samples/subsys/audio/sof/boards/intel_adsp_ace15_mtpm.conf:
```diff
@@ -0,0 +1,11 @@
+CONFIG_INTEL_SSP=y
+CONFIG_INTEL_ALH=y
+CONFIG_LP_MEMORY_BANKS=1
+CONFIG_HP_MEMORY_BANKS=22
```
These aren't Zephyr kconfigs, and SOF really shouldn't be touching the SRAM power control anymore.  If this is actually needed, we probably have some code in SOF that needs to be disabled?  Or maybe it's just some vestigial stuff copied from elsewhere?

kv2019i replied:
@andyross Good catch. Support for getting these via dt was added recently -> "soc/intel_adsp: fetch HPSRAM bank count from devicetree" FYI @aborisovich

aborisovich replied:
Unfortunately this is not applicable to ACE platforms.
We have a 2816K SRAM memory and it does divide to this fixed SRAM_BANK_SIZE at all.
In fact we have a 8 MB virtual addressing to access HPSRAM and dividing total memory by this number does not give you EBBs count. We need to figure out something alternative here,

aborisovich replied:
Done.

andyross reviewed (changes requested):
Sorry, I saw this land while we were all at the conference but am just now getting a chance to review.  And, as this is in the magical handful of hours where I find myself unemployed, I'm maybe being a bit stickier than I normally would as I'm representing no ones interests but my own.  Apologies if anything seems brusque.

Most of the comments are nitpicks, but the blockers to me are:

1. The seemingly dead runtime PM code is a red flag.  We really don't want dead code in the tree.  If it works it should work to run the test suite with CONFIG_PM=y, so we should turn that on for at least some tests.  If the only exercisers are out of tree, we should commit a unit test here that at least covers all the code.  If it's not quite ready we should wait until it is before merging. (Or if I'm missing something just point me to where this gets tested, of course.  Pretty sure the existing PM tests don't run on cAVS/ACE yet?)

2. The hundreds of lines of cut and paste are a hard no from me.  I know exactly how it happened and why it's going to be difficult to work backwards to un-paste the layers.  But this code really was written to be a unified platform and there's no reason we can't have that.

andyross commented:
Oh, and because I know there is organizational urgency here but no longer have access to the relevant organization's communication methods: if you need to find me with low latency, remember I'm almost always available on the Zephyr discord (post-ZDS I even have it working on my phone, thanks to @laurenmurphyx64 for the push there).  The #sof channel is probably a great spot for this stuff.

aborisovich commented on soc/xtensa/intel_adsp/ace_v1x/ace-ipc-regs.h:
```diff
@@ -0,0 +1,119 @@
+#ifndef ZEPHYR_SOC_INTEL_ADSP_ACE_IPC_REGS_H
+#define ZEPHYR_SOC_INTEL_ADSP_ACE_IPC_REGS_H
+
+/* Inter Processor Communication: a distressingly heavyweight method
```
This file has passed PR in out internal process and this is not the place to talk about things that are integrated already.
This is upstream, dropping sorry,

aborisovich replied:
Andy IPC/IDC PR is still in review in the innersource repos and maintainers are encouraging me to rename driver from cavs_ipc to more common name. It makes sense as cAVS and ACE despite their similarities are different SoCs.
I would like to leave this header as it currently is. So we would have something like intel_adsp_ipc driver and different headers for ACE and CAVS IPC registers. To further support my claim, here are arguments for doing so:

1. Splitting struct cavs_ipc for another \#ifdef MTL "case" is not elegant. They are not that similar due to "memory gaps" in MTL registers and would look awful and confusing. Like so:
![image](https://user-images.githubusercontent.com/12482515/174908699-e7afe70b-b54e-46d9-8827-0def34e250f4.png)

1. Putting registers of one SoC family into registers of other SoC family is simply wrong despite its similarities.
2. Why ACE code should know what this definition is:
`#define CAVS_IPC_IDD15_DONE BIT(30) /* v1.5 way to signal done */`

1. Why would CAVS code should know what this definition is:
`#define ACE_IPC_TDA1X_DONE BIT(0)`
(this one is not present here it is under review on innersource but required by ACE).

So if you like objectively look at those two headers and compare their content you will see that like 50% is different.
It is a lot (skipping argumentation that new SoC should have separate definitions).

andyross replied:
I'm not seeing your point.  The differences between the two layouts are that one register (CAP) was added to the end of the block and two registers (IDD and TDD) moved to a spot a few dozen words later in memory.  That seems like pretty straightforward migration churn to me and not worth a separate API.

But really the concern is more: there's a seven[1] paragraph comment here explaining how this interface works, which is clearly applicable to both cAVS and MTL/ACE because it's repeated verbatim in both places.  And if we have an interface so complicated that it requires seven paragraphs to explain and understand[2], we really don't want to be maintaining two different drivers for it.

[1] Good grief, the author was a windbag.

[2] Which, OK, the author is famously a windbag and maybe this isn't so necessary.

aborisovich commented on soc/xtensa/intel_adsp/ace_v1x/ace_v1x-regs.h:
```diff
@@ -0,0 +1,252 @@
+#define MTL_PWRBOOT_BATTR_LPSCTL_L1_MIN_WAY       BIT(15)
+#define MTL_PWRBOOT_BATTR_LPSCTL_BATTR_SLAVE_CORE BIT(16)
+
+#define MTL_PWRBOOT (*(volatile struct mtl_pwrboot *)0x00178d00)
```
This is not a place for refactoring, this is upstream.

carlocaione replied:
well, if @andyross doesn't want to NACK this. I'll do.

NACK, this must be fixed before this goes upstream.

aborisovich replied:
@carlocaione I will like to explain you the context.
Of course issues must be fixed before this goes upstream.
I am talking to @andyross in context of our Intel internal work model.
We have internal repository from which we upstream this code and the work model was to perform all possible fixes and refactorings there so the upstream contains only issues/remarks from external maintainers.
Andy had been a reviewer (and maintainer?) of this internal repository for a long time and he did not file issues there or refactoring requests. He did not block eg. this registry changes that are already integrated work in the inner repository while he should if those changes are not ok in his opinion.
That is why i wrote

> This is not a place for refactoring, this is upstream.

This is in context of me being angry that this work had not been done where it supposed to be.
Of course I will address and fix all issues now, it's my responsibility and only choice now,

carlocaione commented on soc/xtensa/intel_adsp/ace_v1x/linker.ld:
```diff
@@ -0,0 +1,14 @@
+
+MEMORY
+{
+    LINKER_DT_RESERVED_MEM_REGIONS()
```
Too late for this. This was dropped with #45030. Either do what suggested by @andyross or `zephyr,memory-region` can be used for this.

carlocaione reviewed (dismissed):
something to change.

Now, as a side note, this:
> If you wish to refactor you are too late there was a time for it in our internal development repositories.
Further refactorings will be done after upstream process is complete as separate PRs.

is seriously wrong, not how open-source works (because, you know, not everyone has access to your internal repo) and definitely don't put maintainers in a good mood when reviewing your PR.

andyross commented:
So, obviously there's likely to be some strife here and I really do recognize that this isn't my core responsibility anymore.  I'm not trying to be an insurmountable roadblock.  But just to point out something regarding Github UI: the "Resolve conversation" button doesn't actually do anything to accelerate review, all it does is hide the argument.  What you need to do if you need to force the issue is get someone with merge privileges (e.g. @nashif or @MaureenHelm) to dismiss my review and merge it over my objections.  And that kind of thing does happen occasionally and it's not wrong when it does.

But before we get there, I really do strongly suggest we work at cleaning this up a little.  None of this should be all that much work:

1. I recommend splitting out the seemingly-dead PM integration and submitting as a separate PR with test integration so we can at least be sure it builds in-tree (the project can accept on faith that the internal simulation and hardware work produces passing results -- obviously none of us have access).  This shouldn't be a big deal, basically everything has to be inside a `#ifdef CONFIG_PM` already, so it's easy to find.

2. The cut and paste shouldn't be that hard to unify.  I don't see a lot of deltas from the original unified version, which (while I can't point you to it anymore) is still in a not-yet-outrageously-stale topic branch and would be useful as a guide.  The core point is that when originally done, the ACE integration was unified with cAVS and consisted of a few new files and something like 90 lines of changes to existing files.  This PR is adding almost 2000 lines because of all the duplication.

3. Remove the rimage and hal/machine files and resubmit to the proper projects.  That part is trivial (though as mentioned, IIRC rimage actually has that file already I think?).

And please reach out if you need help.  Next week is likely to be chaotic for me, but I'm sure I can find time if needed.

aborisovich commented:
> something to change.
> 
> Now, as a side note, this:
> 
> > If you wish to refactor you are too late there was a time for it in our internal development repositories.
> > Further refactorings will be done after upstream process is complete as separate PRs.
> 
> is seriously wrong, not how open-source works (because, you know, not everyone has access to your internal repo) and definitely don't put maintainers in a good mood when reviewing your PR.

Of course issues must be fixed before this goes upstream and I should not have wrote this part you quoted in the title
(it does not make any sense - that's for review process is).
@carlocaione I would like to explain you the context.
I was talking to @andyross in context of our Intel internal work model.
We have internal repository from which we upstream this code and the work model was to perform all possible fixes and refactorings there, so the upstream contains only issues/remarks from external maintainers.
Andy had been a reviewer (and maintainer?) of this internal repository for a long time and he did not file issues there or refactoring requests. He did not block eg. this registry changes that are already integrated work in the inner repository while he should if those changes are not ok in his opinion.
Of course I will address and fix all issues now, it's my responsibility and only choice now.

andyross commented:
> Andy had been a reviewer (and maintainer?) of this internal repository for a long time and he did not file issues there or refactoring requests.

It's... complicated.  I'm also the original author of most of this PR, and it was originally written as a set of patches on top of upstream, using a single framework.  When it got broken out for the benefit of our internal tree structure, I actually was promised that this was a temporary state and that everything would be cleaned up before upstreaming.  I don't know if you were made aware of that requirement or not, and I'm sure it's frustrating that you're being asked to manage upstreaming for someone else's code.  But there are/were only so many fights I can engage in at a time.

But really none of that matters anymore.  I'm just a reviewer here, and I personally think (with at least some pretty reasonable domain expertise) that the structure of this PR needs to change before it goes in.  I really don't think that should be such a big deal, and I'm absolutely willing to help.

andyross commented on boards/xtensa/intel_adsp_ace15_mtpm/intel_adsp_ace15_mtpm_defconfig:
```diff
@@ -0,0 +1,15 @@
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
Looks like this got added by a script or something?  The kconfig tool uses Bourne shell comments ("#" lines) and doesn't invoke the C preprocessor.  These lines get flagged with a "malformed" warning by kconfig which then aborts the build.

andyross commented on soc/xtensa/intel_adsp/ace_v1x/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,78 @@
+	default 0
+
+config MTL_TIMER
+	default y
```
This produces a warning:

```
warning: MTL_TIMER (defined at soc/xtensa/intel_adsp/ace_v1x/Kconfig.defconfig.series:58) defined without a type
```

Because nothing ever defines the MTL_TIMER kconfig.  Presumably that got dropped from the submission?  The kernel timer driver is a pretty core thing and needs to land at the same time as the platform in most cases.

Also: note that this looks like another cut-and-paste job incoming.  The cAVS and ACE/MTL hardware have identical timer interfaces modulo the address of the registers. (The all-purpose/miscellaneous "shim" block that used to be there got split up into separate blocks in ACE).  They really want to be using the same driver, not a copied version.

marc-hb replied:
Looks like these are being caught by twister in CI too; no need to fetch and build locally:

https://github.com/zephyrproject-rtos/zephyr/runs/6843591805?check_suite_focus=true (RED)

```
-- Generated device_extern.h: /__w/zephyr/zephyr/twister-out/nxp_adsp_imx8m/samples/subsys/audio/sof/sample.audio.sof/zephyr/include/generated/device_extern.h
-- Including generated dts.cmake file: /__w/zephyr/zephyr/twister-out/nxp_adsp_imx8m/samples/subsys/audio/sof/sample.audio.sof/zephyr/dts.cmake

warning: SMP_BOOT_DELAY (defined at kernel/Kconfig:859) was assigned the value 'y' but got the value
'n'. Check these unsatisfied dependencies: SMP (=n). See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_SMP_BOOT_DELAY and/or look up
SMP_BOOT_DELAY in the menuconfig/guiconfig interface. The Application Development Primer, Setting
Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be helpful
too.

warning: MTL_TIMER (defined at soc/xtensa/intel_adsp/ace_v1x/Kconfig.defconfig.series:58) defined without a type
Parsing /__w/zephyr/zephyr/Kconfig
Loaded configuration '/__w/zephyr/zephyr/boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m_defconfig'
Merged configuration '/__w/zephyr/zephyr/samples/subsys/audio/sof/prj.conf'
Merged configuration '/__w/zephyr/zephyr/samples/subsys/audio/sof/boards/nxp_adsp_imx8m.conf'
error: Aborting due to Kconfig warnings
CMake Error at /__w/zephyr/zephyr/cmake/modules/kconfig.cmake:290 (message):
  command failed with return code: 1
Call Stack (most recent call first):
  /__w/zephyr/zephyr/cmake/modules/zephyr_default.cmake:121 (include)
  /__w/zephyr/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:51 (include)
  /__w/zephyr/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:76 (include_boilerplate)
  CMakeLists.txt:5 (find_package)
-- Configuring incomplete, errors occurred!
```

>  I'm sure it's frustrating that you're being asked to manage upstreaming for someone else's code

Ouch, what could possibly go wrong...

andyross reviewed (changes requested):
Pulled it to see if I could find some low hanging fruit to help with and ran into some build failures.

aborisovich commented:
Removed config MTL_TIMER causing build issues.
@andyross adding timer driver here would require adding also interrupt driver. Can we try without it here and upstream it in the second PR?

nashif commented:
> Removed config MTL_TIMER causing build issues.

Why not add the driver or use the existing driver in the tree and modify it for mtl?

> @andyross adding timer driver here would require adding also interrupt driver. Can we try without it here and upstream it in the second PR?

why not add in this PR? How is this ever going to work without timer and interrupt controller hooked in?

ceolin commented on soc/xtensa/intel_adsp/ace_v1x/multiprocessing.c:
```diff
@@ -0,0 +1,205 @@
+		__fallthrough;
+	case PM_STATE_RUNTIME_IDLE:/* D0 */
+		k_cpu_idle();
+		break;
```
we can go even further, if the target just implement `PM_STATE_SOFT_OFF` it does not need to declare other states in DT and handle them here. The idle thread will automatically call `k_cpu_idle()`. My understanding is that there is a desire to map all states that the target supports even if the implementation is not really doing anything special. It is more complicated than it needs to be, but it is not a blocker for me.

tmleman replied:
@andyross @ceolin I don't understand why it's a problem now, since these states have to be implemented anyway. Of course you can do it as you suggested and add all the code for each state separately.

ceolin replied:
It is not a problem for me. I just said that this can be simpler.

tmleman commented on soc/xtensa/intel_adsp/common/ace/asm_memory_management.h:
```diff
@@ -0,0 +1,76 @@
+#define EBB_SEGMENT_SIZE	  32
+#define PLATFORM_HPSRAM_EBB_COUNT 22
+
+.macro m_ace_hpsram_power_change segment_index, mask, ax, ay, az, au, aw
```
@andyross those macros disable lp/hp sram. And since we disable memory we want ace_power_down to be a single function that is executed before shut down of a core.

IMR is slow and we don't wanna be slow.

tmleman commented on soc/xtensa/intel_adsp/common/ace/power_down.S:
```diff
@@ -0,0 +1,159 @@
+#define temp_reg6                    a12
+#define p_ipc_regs                   a13
+#define u32_ipc_response_mask        a14
+#define pfl_reg                      a15
```
@andyross cAVS does it in the same way but code will not be the same. Is not even the same on the all cAVS platforms.

I'm not trill on the idea of a rewrite the entire code/flow at this moment. Especially if it is tested and it works stably.

tmleman commented on soc/xtensa/intel_adsp/ace_v1x/multiprocessing.c:
```diff
@@ -0,0 +1,205 @@
+	}
+}
+
+int soc_adsp_halt_cpu(int id)
```
@andyross I talked about this with @ceolin. I also asked You in April about it, when the initial implementation was made.  

I considered renaming so that it would not be the same function as in cAVS. But at the moment i'm more in favourite of changing cAVS implementation. For me it's misunderstanding when you are intending to halt core but in reality only allowing HW to PG in idle.

andyross commented:
I guess it's feasible to push this without the cavs_timer integration, since there's a fallback XTENSA_TIMER driver based on CCOUNT/CCOMPARE (not SMP-compatible though -- the clocks aren't synchronized).  But that seems like wasted effort to integrate to me.  Really though, I'm less picky about the code that doesn't appear in the PR than the code that does.

My personal blockers remain the cut/paste files and the HAL headers (and rimage config) that needs to be submitted there instead, and I'd really like to see CONFIG_PM turned on so we know the power management code isn't dead.

ceolin commented:
> I guess it's feasible to push this without the cavs_timer integration, since there's a fallback XTENSA_TIMER driver based on CCOUNT/CCOMPARE (not SMP-compatible though -- the clocks aren't synchronized). But that seems like wasted effort to integrate to me. Really though, I'm less picky about the code that doesn't appear in the PR than the code that does.
> 
> My personal blockers remain the cut/paste files and the HAL headers (and rimage config) that needs to be submitted there instead, and I'd really like to see CONFIG_PM turned on so we know the power management code isn't dead.

`CONFIG_PM` code is not dead. This is an optional feature, as I mentioned the majority of boards supporting PM don't enable it by default.  Of course, you can just remove the `ifdef` guard and let the compiler checks the code. But regarding this option not being enabled, that is not a problem.

aborisovich commented:
Thank you all reviewers for help. I will start answering reviews related to code as soon as will get all greens with CI.
Meanwhile I will be very grateful for help related with existing CI failures.
Last updates:

- Fixed most syntax errors from CI step "Compliance Checks / Run compliance checks on patch series (PR) (pull_request)".

Still TODO:

- Remained only false-positives from checkpatch trying to fix assembly code (maintainers - how do we fix it or ignore?) in "Compliance Checks / Run compliance checks on patch series (PR) (pull_request)".
- `CONFIG_METEORLAKE`  in _samples/subsys/audio/sof/boards/intel_adsp_ace15_mtpm.conf:1_
is reported as undefined Kconfig symbol. Question - how those symbols are defined for CAVS platforms _samples/subsys/audio/sof/boards/intel_adsp_cavs25.conf_ has `CONFIG_TIGERLAKE=y` _samples/subsys/audio/sof/boards/intel_adsp_cavs20.conf_ has `CONFIG_ICELAKE=y`
etc.

- _warning: BOUNDS_CHECK_BYPASS_MITIGATION (defined at kernel/Kconfig:770) was assigned the value 'y'
but got the value 'n'. Check these unsatisfied dependencies: USERSPACE (=n)._ Should I add CONFIG_USERSPACE=n
to SoC definition?

- warning: DAI_SSP_HAS_POWER_CONTROL (defined at drivers/dai/intel/ssp/Kconfig.ssp:16) was assigned 
the value 'y' but got the value 'n'. Check these unsatisfied dependencies: DAI_INTEL_SSP (=n), DAI (=n).
I temporary put `CONFIG_DAI_SSP_HAS_POWER_CONTROL=n` to _boards/xtensa/intel_adsp_ace15_mtpm/intel_adsp_ace15_mtpm_defconfig_ but this does not resolved the issue.
DAI_INTEL_SSP resolves to True if dtsi file has intel,ssp-dai compatible node but _dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi_ file does not define such node. I do not understand what is happening with this Twister build exactly?

- CMake gives me errors in CI step "Run tests with twister / twister-build (2) (pull_request_target)":
CMake Error at ../../../cmake/modules/extensions.cmake:418 (add_library):
_add_library cannot create target "intel_adsp_common" because another target with the same name already exists.  The existing target is a static library created in source directory  "/__w/zephyr/zephyr/soc/xtensa/intel_adsp/common"._
Debugging what happened there.

I am sorry that this PR moved out of draft too quickly and reviewers will receive spam now.
I really did not think it would go that bad as it is now.

nashif commented:
> * `CONFIG_METEORLAKE`  in _samples/subsys/audio/sof/boards/intel_adsp_ace15_mtpm.conf:1_
>   is reported as undefined Kconfig symbol. Question - how those symbols are defined for CAVS platforms _samples/subsys/audio/sof/boards/intel_adsp_cavs25.conf_ has `CONFIG_TIGERLAKE=y` _samples/subsys/audio/sof/boards/intel_adsp_cavs20.conf_ has `CONFIG_ICELAKE=y`
>   etc.

does CONFIG_METEORLAKE exist in the SOF tree already? probably not, thats why, I would remove this until SOF is ready for MTL.

nashif commented:
> CMake gives me errors in CI step "Run tests with twister / twister-build (2) (pull_request_target)":
CMake Error at ../../../cmake/modules/extensions.cmake:418 (add_library):
_add_library cannot create target "intel_adsp_common" because another target with the same name already exists. The existing target is a static library created in source directory "/_w/zephyr/zephyr/soc/xtensa/intel_adsp/common".
Debugging what happened there.

You need to remove/adapt the changes in the first commit to `soc/xtensa/intel_adsp/CMakeLists.txt`. Those are not needed now.

nashif commented:
> warning: BOUNDS_CHECK_BYPASS_MITIGATION (defined at kernel/Kconfig:770) was assigned the value 'y'
but got the value 'n'. Check these unsatisfied dependencies: USERSPACE (=n). Should I add CONFIG_USERSPACE=n
to SoC definition?

> warning: DAI_SSP_HAS_POWER_CONTROL (defined at drivers/dai/intel/ssp/Kconfig.ssp:16) was assigned
the value 'y' but got the value 'n'. Check these unsatisfied dependencies: DAI_INTEL_SSP (=n), DAI (=n).
I temporary put CONFIG_DAI_SSP_HAS_POWER_CONTROL=n to boards/xtensa/intel_adsp_ace15_mtpm/intel_adsp_ace15_mtpm_defconfig but this does not resolved the issue.
DAI_INTEL_SSP resolves to True if dtsi file has intel,ssp-dai compatible node but dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi file does not define such node. I do not understand what is happening with this Twister build exactly?


Not sure how you are getting those if the board is not able to build in the first place.

nashif committed:
intel_adsp: meteorlake: add memory driver

Add TLB driver specific for Meteorlake.

Add missing kconfig resource and fix include path
add hpsram power up and ref counter

Use memblock to track physical page usage in mtl tlb driver. The
applications that will be using the tlb memory driver should not
track the physical page mapping to virtual address space:
 - adds an option to use the phys argument value of 0 to tell the
   driver to autonomously assign physical pages upon map request
 - makes the tlb driver use memblock to track mapped physical pages

Co-authored-by: Rafal Redzimski <rafal.f.redzimski@intel.com>
Co-authored-by: Adrian Bonislawski <adrian.bonislawski@intel.com>
Co-authored-by: Marcin Szkudlinski <marcin.szkudlinski@intel.com>
Co-authored-by: Michal Wasko <michal.wasko@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
intel_adsp: meteorlike: add timer driver

Add timer driver based on CAVS driver and adapted for Meteor Lake.

Co-authored-by: Michal Wasko <michal.wasko@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
drivers: Add ACE V1X interrupt controller

Add Designware Interrupt Controller for ACE v1x platform

Co-authored-by: Ederson de Souza <ederson.desouza@intel.com>
Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
drivers: dma: Add gpdma derivative of dw dma for ace

This will add ace compatible DMA driver.

Co-authored-by: Adrian Bonislawski <adrian.bonislawski@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
soc: ace1.x: use common code

Do not duplicate code in new SoC and reuse code from intel_adsp/common.
Move SRAM code into own file in common code and setup SRAM in soc for
MTL platform.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

aborisovich commented:
@nashif @andyross @kv2019i thank you for the help, I resolved those issues and moved forward with CIs.
We are closer now old errors are gone and only one remaining is missing core-isa.h header file for MTL.
It seems like it should be added to SOF to make it work...

kv2019i commented:
@aborisovich wrote:
> old errors are gone and only one remaining is missing core-isa.h header file for MTL. It seems like it should be added to SOF to make it work I will upstream it ASAP.

This should be put to modules/hal/xtensa (@andyross pointed this out, I was thinking this comes via SDK, but that's not the case).

I.e. for cAVS2.5 build, the file is here:
zephyrproject/modules/hal/xtensa/zephyr/soc/intel_cavs_25/xtensa/config/core-isa.h

There should be a folder for ace that has the core config and that is used in the build.

aborisovich commented:
> > @nashif @andyross @kv2019i thank you for the help, I resolved those issues and moved forward with CIs.
> > We are closer now old errors are gone and only one remaining is missing core-isa.h header file for MTL.
> > It seems like it should be added to SOF to make it work I will upstream it ASAP.
> 
> not sure how you are testing, but this board needs to build without SOF in zephyr, just try
> 
> `west build -b intel_adsp_ace15_mtpm samples/hello_world/`
> 
> Fixing this in SOF is not going to help, there is something wrong with the code in this PR.

After moving xtensa headers to zephyrproject/modules/hal/xtensa/zephyr/soc/intel_ace15_mtpm locally as @kv2019i  suggested:
![image](https://user-images.githubusercontent.com/12482515/175013784-fb870973-be39-454f-a805-55d923adb476.png)
and setting export ZEPHYR_BASE=/home/aborisov/workspace/converged1/zephyr-intel where my PR is,
I am getting successful compilation using command:
`west build -b intel_adsp_ace15_mtpm samples/hello_world/ -p always`
Only last step "Linking C executable zephyr/zephyr_pre0.elf" fails and this gets hard to debug (weird syntax error).
Adding logs
[mtl-build-hello-world.log](https://github.com/zephyrproject-rtos/zephyr/files/8956987/mtl-build-hello-world.log)

aborisovich commented:
I've created [PR15](https://github.com/zephyrproject-rtos/hal_xtensa/pull/15) with upstream of that xtensa headers to hal module.

nashif commented:
> I've created [PR15](https://github.com/zephyrproject-rtos/hal_xtensa/pull/15) with upstream of that xtensa headers to hal_xtensa module.

you will need to link that PR in here via a manifest change, to get CI to use this change and thus resolve the issue.

Interesting however that yesterday in my attempt to get all this working, I got the mtl board to build successfully  without needing this.

Commit:
manifest: hal_xtensa: updates to new ace family definitions

Updated hal_xtensa module with new xtensa definitions
for intel_adsp_ace1x soc family.

Signed-off-by: Andrey Borisovich <andrey.borisovich@intel.com>

marc-hb commented on west.yml:
```diff
@@ -143,7 +143,8 @@ manifest:
         - hal
     - name: hal_xtensa
-      revision: 0e577021bb66e644afd067cd9f7c71ab11b62b3d
+      url: https://github.com/aborisovich/hal_xtensa.git
```
@nashif can you elaborate? On one hand `pull/15/head` is readable. On the other hand the SHA1 is deterministic. I think both can be achieved with:
```
 - name: hal_xtensa
      # https://github.com/zephyrproject-rtos/hal_xtensa/pull/15
      revision: 0e577021bb66e644afd067cd9f7c71ab11b62b3d
      url: https://github.com/aborisovich/hal_xtensa.git
```

dcpleung replied:
Manifest changes is always a two-part process. First, set it to the manifest PR so CI can run through it. Once CI passes, the manifest PR can be merged. Then the manifest needs to be updated to point to the SHA. And you let CI run through it again (this time with the actual SHA).

marc-hb replied:
I understand it's made of two steps, I'm wondering what's wrong with using (volatile) SHA1s for the first step too.

dcpleung replied:
Since the PR has not been merged, normal git fetch/pull won't download that commit, and thus git checkout will fail later. This is due to that SHA not in the tree of the fetched/pulled branch (as west simply does fetch only).

nashif replied:
https://docs.zephyrproject.org/latest/develop/modules.html?highlight=module#process-for-submitting-a-new-module

marc-hb replied:
First, sorry I misread the diff (the diff on the main, "conversation" page is incomplete). I thought a `revision: SHA1` was being submitted.

> Since the PR has not been merged, normal git fetch/pull won't download that commit, and thus git checkout will fail later.

With Github it works as long as the SHA1 is 40 digits long and in the fork (github shared storage). It's magic, demo:

```diff
--- a/west.yml
+++ b/west.yml
@@ -143,7 +143,8 @@ manifest:
       groups:
         - hal
     - name: hal_xtensa
-      revision: 0e577021bb66e644afd067cd9f7c71ab11b62b3d
+      # https://github.com/aborisovich/hal_xtensa/commits/add-soc-overlay-for-intel_ace15_mtpm
+      revision: cac06cfebfd36a6f5cf05ca13774c36380ff1dce
       path: modules/hal/xtensa
       groups:
         - hal
```

```shell
west update -n hal_xtensa
=== updating hal_xtensa (modules/hal/xtensa):
--- hal_xtensa: fetching, need revision cac06cfebfd36a6f5cf05ca13774c36380ff1dce
remote: Enumerating objects: 14, done.
remote: Counting objects: 100% (14/14), done.
remote: Compressing objects: 100% (10/10), done.
remote: Total 14 (delta 2), reused 14 (delta 2), pack-reused 0
Unpacking objects: 100% (14/14), 22.67 KiB | 7.56 MiB/s, done.
From https://github.com/zephyrproject-rtos/hal_xtensa
 * branch                      cac06cfebfd36a6f5cf05ca13774c36380ff1dce -> FETCH_HEAD
M	zephyr/soc/intel_apl_adsp/xtensa/config/core-isa.h
Previous HEAD position was 0e577021bb66 zephyr: add SoC overlay for NXP i.MX8MP
HEAD is now at cac06cfebfd3 zephyr: add SoC overlay for intel_ace15_mtpm
```

Check https://github.com/thesofproject/rimage/pull/71 for more details.

> This clearly links to the PR and shows that a PR needs to be reviewed and merged first.

Thx.

aborisovich commented:
Push update:

- put proper revision to west.yml for hal_xtensa
- squashed commit [samples: sof: add intel_adsp_mtl configuration](https://github.com/zephyrproject-rtos/zephyr/pull/46323/commits/338368c8ca91171eb95ed6bc297d1bb8663774a7) with [intel_adsp: add the board definition for the Meteorlake](https://github.com/zephyrproject-rtos/zephyr/pull/46323/commits/f1426e1981be0d66220e262417356a2f2ee1e481) as @mwasko suggested.
- cherry-picked changes prepared by @nashif that **saved the day**:

1. added new drivers for memory and IMR, interrupt controller, timer, gpdma
2. squashed new @rfredzim PR "Corrections to MTL memory management" with memory driver commit
3. split linker script adding ace-link.ld 
4. replaced `processor_id_reg()` with `arch_proc_id()` from [PR46659](https://github.com/zephyrproject-rtos/zephyr/pull/46659)
5. Added macro `HPSRAM_REGS` to _soc/xtensa/intel_adsp/ace_v1x/ace_v1x-regs.h_ which was defined only in SOF code

aborisovich commented:
Push update:

- fixed checkpatch syntax issues
- added drivers/dma/Kconfig.ace_gpdma with missing Kconfig symbol CONFIG_DMA_ACE_GPDMA
- fixed invalid usage of __ASSERT macro causing failures in twister tests

TODO:

- Add missing CONFIG_MM_DRV_INTEL_ADSP_IMR_MEMORYBLOCK symbol with proper description

nashif commented on soc/xtensa/intel_adsp/ace_v1x/mtl-regs.h:
```diff
@@ -0,0 +1,12 @@
+#ifndef _ZEPHYR_SOC_INTEL_ADSP_MTL_REGS_H_
+#define _ZEPHYR_SOC_INTEL_ADSP_MTL_REGS_H_
+
+/* mtl-regs.h is deprecated, use ace_v1x-regs.h instead */
```
@aborisovich  you need to remove this header, it is not being used anywhere.

aborisovich replied:
It's used in SOFs platform.h but we will get rid of it. Removed.

kv2019i reviewed (dismissed):
I know this is still marked as draft and some tests are failing, but I'll record here that all my blocking concerns have been addressed now. Thanks @aborisovich !

aborisovich commented on soc/xtensa/intel_adsp/ace_v1x/mtl.toml:
```diff
@@ -0,0 +1,57 @@
+version = [1, 8]
+
```
Removed.

aborisovich commented:
Push update:

- Removed mtl.toml as not needed here (it is used by rimage and is placed in rimage repository)
- Added missing CONFIG_DAI_SSP_HAS_POWER_CONTROL=y to board configuration
- Cherry-picked @nashif commits:

1. with code refactorings to avoid code duplication in new SoC and reuse code from intel_adsp/common
2. fixing configuration/building of mm driver
- Removed obsolete mtl-regs.h header

andyross commented on drivers/mm/mm_drv_intel_adsp_imr_memoryblock.c:
```diff
@@ -0,0 +1,83 @@
+	((uintptr_t)(addr) >= (uintptr_t)IMR_BASE_ADDR &&		\
+	(uintptr_t)(addr) <  (uintptr_t)(IMR_BASE_ADDR + IMR_MEMORY_SIZE))
+
+int intel_adsp_ddr_memory_get(void *address, size_t length)
```
Style: the API in this file is a **really** thin wrapper around the pre-existing sys_mem_blocks API.  Maybe just use that directly?  What's the value being added here, because needless abstractions have cost.

andyross commented on drivers/timer/ace_v1x_timer.c:
```diff
@@ -0,0 +1,224 @@
+# define COUNTER_LO (&CAVS_SHIM.dspwc_lo)
+# define COMPARE_HI (&CAVS_SHIM.UTIL_CAT(UTIL_CAT(dspwct, COMPARATOR_IDX), c_hi))
+# define COMPARE_LO (&CAVS_SHIM.UTIL_CAT(UTIL_CAT(dspwct, COMPARATOR_IDX), c_lo))
+#endif
```
Still cutting and pasting here.  I note (heh) that the blocks above implement drivers for both ACE and cAVS hardware via some slightly kludgey pointer macros, and that the rest of the file looks essentially identical to the pre-existing cavs_timer.c in the same directory.  I'm guessing that if you drop this in place on top of cavs_timer.c and have the MTL board enable CONFIG_CAVS_TIMER (and no doubt fix up a few collisions and bitrot spots), the same driver will work for the whole hardware family.

Shorter version: the HDA wall clock timer register interface didn't change at all in ACE, they just moved out of the crowded "shim" block to a timer-specific "TTS" block.

andyross commented on soc/xtensa/intel_adsp/ace_v1x/soc.c:
```diff
@@ -0,0 +1,82 @@
+	int prid;
+	__asm__ volatile("rsr %0, PRID" : "=r"(prid));
+	if (prid != 0) {
+		((void(*)(void))MTL_PWRBOOT.bootctl[prid].baddr)();
```
This workaround should still come out if we aren't going to use the BADDR mechanism (though IMHO we should, it's much cleaner than the lpsram trampoline trick).

andyross reviewed (dismissed):
Swing by to review again.  Definitely looking better to me.  Still don't like the duplicated IPC header (though the driver is still to come I guess).  Also the timer driver is here now, which is good, but it's another cut/paste job and needs to be unified with cavs_timer (which AFAICT it already is, the cavs code is in this version!).

aborisovich commented:
Push update:

- fixed duplicated namings in idelay function
- changed sram1 node size to 64KB according to L2 ULP Memory size in documentation
- pulled in changes fixing timer usage for MTL for common/kernel application

TODO:

- in progress -> moving hardcoded address definitions from soc/xtensa/intel_adsp/ace_v1x/ace_v1x-regs.h to devicetree

Commit:
soc: ace1x: provided register names to addresses

Provided register names to hardcoded addresses in ace_v1x-regs
header.

Signed-off-by: Andrey Borisovich <andrey.borisovich@intel.com>

aborisovich commented:
Push update:
refactored soc/xtensa/intel_adsp/ace_v1x/ace-ipc-regs.h file:
- removed unused MTL_SB_IPC macro
- reduced verbosity of remarks about IPC and IDC
- provided register name to hardcoded address
- provided fancy doxygen syntax

@andyross Is this file now bearable?
Please note that IPC/IDC driver will be coming and is not part if this PR.

dcpleung commented on drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:
```diff
@@ -0,0 +1,619 @@
+ * Number of significant bits in the page index (defines the size of
+ * the table)
+ */
+#if defined(CONFIG_SOC_SERIES_INTEL_CAVS_V15)
+# define TLB_PADDR_SIZE 9
```
This is a MTL driver so there is no need for CAVS stuff here.

dcpleung commented on drivers/mm/mm_drv_intel_adsp_imr_memoryblock.c:
```diff
@@ -0,0 +1,83 @@
+#define IMR_NUM_OF_PAGES (IMR_MEMORY_SIZE / IMR_PAGE_SIZE)
+
+/* declare an IMR memory block */
+SYS_MEM_BLOCKS_DEFINE_WITH_EXT_BUF(
+		IMR_REGION,
+		IMR_PAGE_SIZE,
+		IMR_NUM_OF_PAGES,
+		(uint8_t *) IMR_BASE_ADDR);
```
Does this driver get built at all? Compiler should have complained about duplicate identifier `IMR_REGION` because the mem_block is named `IMR_REGION` and you also have a `#define IMR_REGION` above.

Commit:
intel_adsp: Unify cavs and ace timers

These two timers were sharing pretty much the same code. Actually
mtl timer was a "superset" of cavs timer. Just merge them into a
single one called intel audio dsp timer (intel_adsp_timer).

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

dcpleung commented:
BTW, there are a bunch of compliance check failures that need to be addressed. Most of them are white space errors.

aborisovich commented on samples/subsys/audio/sof/prj.conf:
```diff
@@ -21,3 +21,5 @@ CONFIG_SMP_BOOT_DELAY=y
 # Fix the sys ticks value  until following bugs are solved:
 # - https://github.com/zephyrproject-rtos/zephyr/issues/46378
 CONFIG_SYS_CLOCK_TICKS_PER_SEC=15000
+CONFIG_DAI=y
+CONFIG_HEAP_MEM_POOL_SIZE=2048
```
Yeah, twister runs building SOF application for every cavs platform so if they are green this must be working.
We turned of ace board from twister because we can't resolve Kconfig dependencies pretty well but rest of the platforms are fine.

dcpleung replied:
The GitHub CI does not run anything on audio hardware. Have you tested it to make sure this doesn't break non-MTL platforms? If not, I recommend only enabling these kconfigs in board overlay.

aborisovich commented:
> BTW, there are a bunch of compliance check failures that need to be addressed. Most of them are white space errors.

All compliance checks were fixed.
Those that are left are false-positives related to checkpatch scanning assembly file.
I added some new commits now and there may be something new but I keep that monitored :-)

Commit:
soc: ace: Move power management to its own file

Move power management hooks to its own C file.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

aborisovich commented:
Push update:
cherry-picked refactorings prepared by @ceolin:

- As @andyross suggested, moved power functions from multiprocessing.c to own power.c file
- Unified timer driver by moving common code to intel_adsp_timer.c
Old Kconfig option config CAVS_TIMER had to stay as alias to new value due to source code in SOF that uses it.
Once we remove it in SOF we can remove it here.

Commit:
soc: ace: Organizing include headers

Properly prefixing some include files with "zephyr", also organizing
the order they are included.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
soc: adsp: sram: Assorted cosmetic fixes

General code style fixes.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

aborisovich commented:
Push update:

- fixed code styling problems
- @nashif resolved problems with ace soc [hal_xtensa PR15](https://github.com/zephyrproject-rtos/hal_xtensa/pull/15)  including [tie-asm-cavs25.h](https://github.com/zephyrproject-rtos/hal_xtensa/pull/15/files/cac06cfebfd36a6f5cf05ca13774c36380ff1dce#diff-555a206cfbea7ce630ac5e1a565d0b2303fa1dc9f65d66eb2adc32fafd4309bf) header needlessly
- [hal_xtensa PR15](https://github.com/zephyrproject-rtos/hal_xtensa/pull/15) had been integrated and merged
- updated hal_xtensa module SHA to upstream commit containing new definitions for intel_adsp_ace1x family
- @ceolin fixed Kconfig issues with timer refactorings causing SOF twister tests to fail for cavs platforms

marc-hb commented:
Testing 52e2f18a54173c9ae77355b203a190038670360c on a _limited_ number of HW models in https://github.com/thesofproject/sof/pull/5980

nashif commented:
> It's not... not good. In [sof-ci.01.org/sofpr/PR5980/build803/devicetest](https://sof-ci.01.org/sofpr/PR5980/build803/devicetest), Zephyr booted on APL but on neither ADL, TGL, CML or WHL.

try now. it was something minor indeed.

andyross reviewed (dismissed):
Thanks so much for the cleanup of the timer driver.

As far as IPC: this does indeed successfully defer the argument, because the full IPC driver isn't included here (the register interface is used, but for the ACE-only sched_ipi() implementation which for performance reasons doesn't want to do the full two-way IPC negotiation and just wants a level triggered interrupt).

So... just be aware there's another -1 lurking from me if the IPC driver arrives in a way that isn't a straightforward evolution of cavs_ipc. :)

I'm sure I had other nitpicks I've forgotten, but really these were the only blockers.  Giving a +1 for posterity, though it looks like a few more revs will be needed.

andyross commented:
Oh, and so folks are aware (since I have no sideband way to communicate this anymore): I'm currently on vacation.  The vestigial ghost of my Intel sabbatical runs from this week until July 23rd.  This week I'm on the beach and available, but the next two I'll be in the Galapagos and much harder to reach.  If you want to merge something without me getting in the way, this is your moment.

Commit:
soc: intel_adsp: reading HP SRAM banks count for ACE1X from Devicetree

Replaced hardcoded for intel_adsp_ace15_mtpm board
HP_MEMORY_BANKS value used in SOF code with generic approach -
using PLATFORM_HPSRAM_EBB_COUNT read from Devicetree.

Signed-off-by: Andrey Borisovich <andrey.borisovich@intel.com>

aborisovich commented:
Push update:
- renamed HSXPGCTL_REG to L2HSBPM_REG to better describe base register address
with its offsets (L2HSBPM_REG is base register family address and HSXPGCTL conctere register
at this offset)
- removed unused CAVS macros from meteorlake memory driver
- renamed config DW_ACE_V1X_ICTL to DW_ICTL_ACE_V1X as @dcpleung suggested
- removed hardcoded CONFIG_HP_MEMORY_BANKS=22 for new ace board and enabled usage of
generic PLATFORM_HPSRAM_EBB_COUNT macro (adressed @kv2019i and @andyross issue).

aborisovich commented:
Push update:
- fixed another checkpatch coding style issues
- Implemented @nashif change that resolves issues with APL failing to boot:
  moved `#define PLATFORM_DISABLE_L2CACHE_AT_BOOT` from _soc/xtensa/intel_adsp/common/sram.c_ to _soc/xtensa/intel_adsp/common/boot.c_ .

marc-hb commented:
> try now. it was something minor indeed.

I did and test results in https://sof-ci.01.org/sofpr/PR5980/build824/devicetest/ look MUCH better. There seems to be some unrelated issue with just one device (sh-adlp-rvp-sdw-02), everything else is fine.

What was the issue?

nashif commented:
> What was the issue?

missing define that would enable sram initialization.

teburd reviewed (commented):
Is there a matching dma_ace_gpdma.c that might've been missed here? I see the dts binding and Kconfig for ace gpdma but the driver doesn't seem to be here

dcpleung commented on drivers/mm/mm_drv_intel_adsp_imr_memoryblock.c:
```diff
@@ -0,0 +1,83 @@
+#include <zephyr/kernel.h>
+#include <zephyr/sys/mem_blocks.h>
+#include <zephyr/linker/linker-defs.h>
+#include <zephyr/drivers/mm/mm_drv_intel_adsp_imr_memoryblock.h>
```
This file is missing from this PR.

dcpleung replied:
If you don't need this driver right now, I suggest removing this and add back later with a fixed version.

dcpleung reviewed (changes requested):
There are errors building `drivers/mm/mm_drv_intel_adsp_imr_memoryblock.c` if corresponding is enabled.

Commit:
soc: intel_adsp: added casting for C++ compiler in soc.h

Header soc.h is included during C++ source file compilation
and required adding C++ casts as implicit casting from void*
is forbidden. Fixed minor warning comparing signed with unsigned
integer.

Signed-off-by: Andrey Borisovich <andrey.borisovich@intel.com>

aborisovich commented:
Push update:
As @dcpleung noticed IMR driver source file was not compiled and had errors when enabled.
Removed IMR driver from this PR.

ceolin commented on soc/xtensa/intel_adsp/common/include/soc.h:
```diff
@@ -136,12 +136,16 @@ static ALWAYS_INLINE void z_idelay(int n)
 {
+	#ifdef __cplusplus
+	uint32_t *d = reinterpret_cast<uint32_t *>(dest);
+	uint32_t *s = reinterpret_cast<uint32_t *>(src);
```
please, use C style casts and remove this ifdef.
```
uint32_t *d = (uint32_t *)dest;
uint32_t *s = (uint32_t *)src;
```

aborisovich replied:
No, I thought about this cast for a while and reinterpret_cast is best solution here.
In case of C++ there is no reason to use inefficient C-cast style.
As a reminder, C++ C-cast style performs different operation than in C.
It does following:
1. tries: const_cast
2. tries: static_cast (which is not desirable in this case as this code is invoked from C and literally anything can be passed as source and destination. Most of the casts would fit to this case however).
3. tries static_cast , then const_cast (change type + remove const)
4. tries reinterpret_cast (the one I used here).
5. tries reinterpret_cast , then const_cast (change type + remove const)

Reference: https://en.cppreference.com/w/cpp/language/explicit_cast

To sum up:
`uint32_t *d = (uint32_t *)dest;`
introduced overhead of:
99% of cases - 1 extra cast attempt (const cast)
1% of cases - 3 extra casts attempts

That being said, C-cast is considered a bad practice in C++.
Some explanation: https://isocpp.org/wiki/faq/coding-standards#coding-stds-cpp-vs-c
Some more explanation: https://isocpp.org/wiki/faq/coding-standards#pointer-casts

marc-hb replied:
I understand writing C code in C++ is bad and that C-style casts are a complicated mess in C++ but does any of this "overhead" happen at run time? I don't see `dynamic_cast` in this list. In other words, does this make any difference to the generated binary? I just want to make sure I understand what you mean by "overhead".

BTW:
- https://github.com/zephyrproject-rtos/zephyr/issues/45785

aborisovich replied:
Ok I will use C-cast for now. C++ style is not implemented anyways and who cares about compile time overhead ;-)

marc-hb replied:
Interestingly, Bjarne Stroustrup recommends `uint32_t *d = static_cast<uint32_t *>(dest);` instead in section 5.6 of his C++ book https://eli.thegreenplace.net/2009/11/16/void-and-casts-in-c-and-c

(I opened the book, found section 5.6 and successfully fact-checked this page)

This is what's great with C++: spoiling us with choice :-P

I wonder whether the real question that the various compilers are trying to ask us is maybe this: why does this function accepts `void *` when it actually supports only `uint32_t *`? This looks like a bug...

aborisovich replied:
Well normally in only C++ clean code we obviously should use static cast in this case as you said.
My concerns are related to interopterability with C where we clearly can pass anything as void* to the function.

nashif commented:
net test failure unrelated,

marc-hb commented:
Most SOF testing is failing today, example: https://sof-ci.01.org/linuxpr/PR3742/build343/devicetest/

(non-Zephyr tests are fine)

marc-hb commented:
and the winner is....
- https://github.com/zephyrproject-rtos/zephyr/pull/47524

Apologies to @ceolin , I would have never expected a one-line CMake change to have such a devastating _run-time_ impact so the only other actual xtensa code change was the only (and wrong) suspect left.
